### DISCUSSION TITLE
How to replace vuex with pinia in the existing project?

### DISCUSSION THREAD

**Discussion by ljubadr** [quasarframework/quasar#11902] [General - Components / Directives / etc] [answered]
In my existing quasar project I'm currently using `vuex` but I would like to switch to `pinia`

I found [this](https://github.com/quasarframework/quasar/discussions/10328) and [this](https://github.com/quasarframework/quasar/discussions/9209) discussion on how to setup `pinia` with boot files. Is there a way to have `pinia` added in `client-entry.js` auto-generated file?

Once I setup `pinia`, how can I remove `vuex` from the project? For example, I see that `.quasar/client-entry.js` file has `app.use(store, storeKey)` line, but this file is auto generated

If possible, I would like to avoid running `quasar create` to generate new project without vuex and then adding pinia

**Comment by pdanpdan** [comment]
I'm on mobile, so cannot write too much.
If you can join the discord channel, search for posts about pinia from Tobias Mesquita in v2-other channel.
It has it pretty nicely explained.
If you can then put a summary in here so others can benefit it would help us.
Thank you

**Comment by ljubadr** [comment, marked as answer]
Thanks @pdanpdan for the help!

I've managed to replace `vuex` with `pinia` by following [these steps](https://discord.com/channels/415874313728688138/856517299816497173/925345634129690625) from @TobyMosque 

I use quasar with typescript

Here are the steps for the new or existing project that already use `vuex` that @TobyMosque suggested
- remove `vuex` with `yarn remove vuex`
- install `pinia` with `yarn add pinia`
- modify your `src/store/index.ts` to use `pinia`. I used link that @TobyMosque provided from `TobyMosque/quasar-v2-ssr-pinia` repo and I've used [src/store/index.ts](https://github.com/TobyMosque/quasar-v2-ssr-pinia/blob/master/src/store/index.ts) as a starting point

I didn't test SSR but I'll use [TobyMosque/quasar-v2-ssr-pinia](https://github.com/TobyMosque/quasar-v2-ssr-pinia) as an example if I need it

I've refactored my `vuex` store and modules into `pinia` and everything works fine

Only issue that I have is with boot file where I make the ajax call and save that in the `pinia` store. I try to save data into store `src/store/user.ts` but I'm not sure how to do that properly. I tried to import store

```js
import { useUserStore } from 'src/store/user'

export default boot(....
```

but this doesn't work and I get following error message

> [Quasar] boot error: Error: [🍍]: getActivePinia was called with no active Pinia. Did you forget to install pinia?

I'm not sure how to access `src/store/user.ts` store in the boot file

For now I moved this init logic into  `App.vue` until I figure this out

**Comment by MilosPaunovic** [comment]
Have you tried using the boot file arguments, as described here: 
[https://quasar.dev/quasar-cli/boot-files#anatomy-of-a-boot-file](https://quasar.dev/quasar-cli/boot-files#anatomy-of-a-boot-file)

```js
// somebootfile.js

export default ({ store }) => {
  // Use imported store in any way you like
}
```

**Reply by ljubadr** [reply]
I saw `store` boot argument but I wasn't sure how to access my store that's defined in `src/store/user.ts`. I tried this

```js
import { useUserStore } from 'src/store/user'

export default ({ store }) => {
  const useUserStore = useUserStore()
}
```

but that was throwing an error

> [Quasar] boot error: Error: [🍍]: getActivePinia was called with no active Pinia. Did you forget to install pinia?

@TobyMosque answered that part

> ... because all boots run before the createApp be called.

and he provided examples on how to use store properly in boot file.

One more important thing that he mentioned for the examples that he provided

> in the above example, u also would avoid do any http call using this axios instance from any boot, since that will trigger the interceptor and create the store before the app de created, and u don't wanna that.:

I'll move my inital ajax call out of boot files 😄

**Comment by TobyMosque** [comment]
Regarding use stores in the boot, take on mind that didn't play well with pinia plugins.:
https://pinia.vuejs.org/core-concepts/plugins.html#introduction

> Plugins are only applied to stores **created after `pinia` is passed to the app**, otherwise they won't be applied.

I'm saying that, because all boots run before the createApp be called.
so, let's imagine u had a plugin who inject some depedencies to the stores and another one who persist the state to a cookie.:

```ts
import { boot } from 'quasar/wrappers';
import { AxiosInstance } from 'axios'
import { Router } from 'vue-router';

declare module 'pinia' {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  export interface DefineStoreOptionsBase<S, Store> {
    // allow defining a number of ms for any of the actions
    useCookies?: boolean;
  }
}

declare module 'pinia' {
  export interface PiniaCustomProperties {
    // by using a setter we can allow both strings and refs
    $api: AxiosInstance;
    $router: Router;
  }
}

export default boot(({ app, router, ssrContext }) => {
  const pinia = store as Pinia;
  // this plugin allow u to access $api and $router from any store (actions/getters)
  pinia.use(({ store }) => {
    store.$api = app.config.globalProperties.$api as AxiosInstance;
    store.$router = app.config.globalProperties.$router as Router;
  });
  
  // this plugin copy the state to a cookie, this will happen everytime the state is modified, aswell, that will restore the state from the cookie when the store is created.
  pinia.use(({ store, options }) => {
    if (options.useCookies) {
      const name = store.$id;
      const values = cookies.getAll();
      const prefix = `${name}|`;
      for (const key in values) {
        if (key.startsWith(prefix)) {
          const property = key.replace(prefix, '');
          store[property] = values[key] as never;
        }
      }
      store.$subscribe((_, state) => {
        discard({ _ })
        const keys = Object.keys(state)
        for (const key of keys.filter(key => key !== 'token')) {
          cookies.set(`${prefix}${key}`, state[key], {
            path: '/',
          });
        }
      });
    }
  })
})
```

but what will happen if u do that?
```ts
import { boot } from 'quasar/wrappers';
import useAppState from 'src/stores/app'
import axios from 'axios';

export default boot(async ({ app, store, ssrContext }) => {
  const pinia = store as Pinia;
  const instance = axios.create({
    baseURL: process.env.API_URL,
  });
  
  // useAppState is being called before the createApp be called, so the previouly created plugins will not me applied to it.
  const appState = useAppState(pinia);
  instance.interceptors.request.use(
    (config) => {
      // since the plugin wasn't be applied to this store, the token will be empty, even if the token as be persisted in the cookie in a previous session
      // aswell, the $api and $route will never be avaliable to that store.
      const token = appState.token.value;
      const headers = config.headers as Record<string, string>;
      if (token) {
        headers.Authorization = `Bearer ${token}`;
      }
      return config;
    },
    (error) => {
      return Promise.reject(error);
    }
  );
})
```

in this case, u would move the `useAppState` from the root to inside the interceptor.

```ts
import { boot } from 'quasar/wrappers';
import useAppState from 'src/stores/app'
import axios from 'axios';

export default boot(async ({ app, store, ssrContext }) => {
  const pinia = store as Pinia;
  const instance = axios.create({
    baseURL: process.env.API_URL,
  });
  
  instance.interceptors.request.use(
    (config) => {
      // useAppState is being called after the createApp be called, so the state will be restored as expected.
      // aswell the $api and $store will be avaliable from inside the store.
      const appState = useAppState(pinia);
      const token = appState.token.value;
      const headers = config.headers as Record<string, string>;
      if (token) {
        headers.Authorization = `Bearer ${token}`;
      }
      return config;
    },
    (error) => {
      return Promise.reject(error);
    }
  );
})
```

in the above example, u also would avoid do any http call using this axios instance from any boot, since that will trigger the interceptor and create the store before the app de created, and u don't wanna that.:

**Reply by ljubadr** [reply]
Thank you for your thorough answer and for the examples! This gives me a better idea on how things work and I'll look into this after work

Is `useAppState(pinia)` just a `pinia` store? I don't see import in the example but I guess it's import for `export default defineStore({`

**Comment by TobyMosque** [comment]
Yes, it's a pinia store defined by the defineStore.

Em qua, 5 de jan de 2022 15:37, ljubadr ***@***.***> escreveu:

**Reply by ljubadr** [reply]
Thanks for all the help!
